patch/attiny88.yaml

```diff
@@ -0,0 +1,127 @@
+        ADC4: [4, "ADC Single Ended Input pin 4"]
+        ADC5: [5, "ADC Single Ended Input pin 5"]
+        ADC6: [6, "ADC Single Ended Input pin 6"]
+        ADC7: [7, "ADC Single Ended Input pin 7"]
+        TEMPSENS: [8, "Temperature sensor"]
+        ADC_VBG: [14, "Internal Reference (VBG)"]
+        ADC_GND: [15, "0V (GND)"]
+    REFS0:
+      _replace_enum:
+        INTERNAL: [0, "Internal 1.1V Voltage Reference"]
+        AVCC: [1, "AVcc Reference"]
+  ADCSRA:
+    ADPS:
+      _replace_enum:
+        PRESCALER_2: [1, "Prescaler Value 2"]
+        PRESCALER_4: [2, "Prescaler Value 4"]
+        PRESCALER_8: [3, "Prescaler Value 8"]
+        PRESCALER_16: [4, "Prescaler Value 16"]
+        PRESCALER_32: [5, "Prescaler Value 32"]
+        PRESCALER_64: [6, "Prescaler Value 64"]
+        PRESCALER_128: [7, "Prescaler Value 128"]
+  ADCSRB:
+    ADTS:
+      _replace_enum:
+        FREE: [0, "Free Running mode"]
+        AC: [1, "Analog Comparator"]
+        INT0: [2, "External Interrupt Request 0"]
+        TC0_CMA: [3, "Timer/Counter0 Compare Match A"]
+        TC0_OVF: [4, "Timer/Counter0 Overflow"]
+        TC1_CMB: [5, "Timer/Counter1 Compare Match B"]
+        TC1_OVF: [6, "Timer/Counter1 Overflow"]
+        TC1_CE: [7, "Timer/Counter1 Capture Event"]
+CPU:
+  _modify:
+    CLKPR:
+      access: read-write
+    OSCCAL:
+      access: read-write
+    PORTCR:
+      access: read-write
+    PRR:
+      access: read-write
+  CLKPR:
+    CLKPS:
+      _replace_enum:
+        PRESCALER_1: [0, "Prescaler Value 1"]
+        PRESCALER_2: [1, "Prescaler Value 2"]
+        PRESCALER_4: [2, "Prescaler Value 4"]
+        PRESCALER_8: [3, "Prescaler Value 8"]
+        PRESCALER_16: [4, "Prescaler Value 16"]
+        PRESCALER_32: [5, "Prescaler Value 32"]
+        PRESCALER_64: [6, "Prescaler Value 64"]
+        PRESCALER_128: [7, "Prescaler Value 128"]
+        PRESCALER_256: [8, "Prescaler Value 256"]
+  SMCR:
+    SM:
+      _replace_enum:
+        IDLE: [0, "Idle"]
+        ADC: [1, "ADC Noise Reduction"]
+        PDOWN: [2, "Power-down"]
+EEPROM:
+  EECR:
+    EEPM:
+      _replace_enum:
+        ATOMIC: [0, "Atomic (erase and write in one operation)"]
+        ERASE: [1, "Erase only"]
+        WRITE: [2, "Write only"]
+EXINT:
+  EICRA:
+    ISC0:
+      _replace_enum:
+        LOW: [0, "The low level of INT0 generates an interrupt request"]
+        TOGGLE: [1, "Any logical change on INT0 generates an interrupt request"]
+        FALLING: [2, "The falling edge of INT0 generates an interrupt request"]
+        RISING: [3, "The rising edge of INT0 generates an interrupt request"]
+    ISC1:
+      _replace_enum:
+        LOW: [0, "The low level of INT1 generates an interrupt request"]
+        TOGGLE: [1, "Any logical change on INT1 generates an interrupt request"]
+        FALLING: [2, "The falling edge of INT1 generates an interrupt request"]
+        RISING: [3, "The rising edge of INT1 generates an interrupt request"]
```

Comment:
I'd bunch the two fields together:

```yaml
ISC?:
  _replace_enum:
      LOW: [0, "The low level of INTx generates an interrupt request"]
      TOGGLE: [1, "Any logical change on INTx generates an interrupt request"]
      FALLING: [2, "The falling edge of INTx generates an interrupt request"]
      RISING: [3, "The rising edge of INTx generates an interrupt request"]
```

Ideally this might also be reusable for the other chips at some point ...

Reply:
> I'd bunch the two fields together

That's how I wrote it at first, but it didn't end up replacing the enum values for ISC1.

... *goes off to look through `svdpatch.py`* ...

I *think* the change to make is this, but I'm really out of my element here:

```patch
diff --git a/patch/svdpatch.py b/patch/svdpatch.py
index c70c3a3..1281e9c 100644
--- a/patch/svdpatch.py
+++ b/patch/svdpatch.py
@@ -629,7 +629,7 @@ def process_field_enum(pname, rtag, fspec, field, usage="read-write"):
     derived = None
     for ftag in iter_fields(rtag, fspec):
         name = ftag.find('name').text
-        if derived is None:
+        if derived is None or replace_if_exists:
             enum = make_enumerated_values(name, field, usage=usage)
             enum_name = enum.find('name').text
             enum_usage = enum.find('usage').text
```

Reply:
I've applied the above in fddd2d8, let me know if that looks right to you.

Reply:
Hmm, I don't think that's quite right.  You're essentially short-circuiting the derived enum mechanism.  Please try this patch instead:

```diff
From 150847c8079e9f4cbd8542051c5e37e9472d5a75 Mon Sep 17 00:00:00 2001
From: Rahix <rahix@rahix.de>
Date: Wed, 5 Aug 2020 06:05:21 +0200
Subject: [PATCH] svdpatch: Fix _replace_enum when used on derived fields

Make sure the enum is replaced properly in the case of derived
enumerated values.

Signed-off-by: Rahix <rahix@rahix.de>
---
 patch/svdpatch.py | 30 +++++++++++++++++-------------
 1 file changed, 17 insertions(+), 13 deletions(-)

diff --git a/patch/svdpatch.py b/patch/svdpatch.py
index 0d1f592be450..7d75c1c9d541 100644
--- a/patch/svdpatch.py
+++ b/patch/svdpatch.py
@@ -632,24 +632,28 @@ def process_field_enum(pname, rtag, fspec, field, usage="read-write"):
         field = field["_replace_enum"]
         replace_if_exists = True
 
-    derived = None
+    derived, enum, enum_name, enum_usage = None, None, None, None
     for ftag in iter_fields(rtag, fspec):
         name = ftag.find('name').text
-        if derived is None:
+
+        if enum is None:
             enum = make_enumerated_values(name, field, usage=usage)
             enum_name = enum.find('name').text
             enum_usage = enum.find('usage').text
-            for ev in ftag.iter('enumeratedValues'):
-                ev_usage_tag = ev.find('usage')
-                ev_usage = ev_usage_tag.text if ev_usage_tag is not None else 'read-write'
-                if ev_usage == enum_usage or ev_usage == "read-write":
-                    if replace_if_exists:
-                        ftag.remove(ev)
-                    else:
-                        print(pname, fspec, field)
-                        raise SvdPatchError(
-                            "{}: field {} already has enumeratedValues for {}. Use '_replace_enum' to overwrite."
-                            .format(pname, name, ev_usage))
+
+        for ev in ftag.iter('enumeratedValues'):
+            ev_usage_tag = ev.find('usage')
+            ev_usage = ev_usage_tag.text if ev_usage_tag is not None else 'read-write'
+            if ev_usage == enum_usage or ev_usage == "read-write":
+                if replace_if_exists:
+                    ftag.remove(ev)
+                else:
+                    print(pname, fspec, field)
+                    raise SvdPatchError(
+                        "{}: field {} already has enumeratedValues for {}. Use '_replace_enum' to overwrite."
+                        .format(pname, name, ev_usage))
+
+        if derived is None:
             ftag.append(enum)
             derived = make_derived_enumerated_values(enum_name)
         else:
-- 
2.27.0
```